[FLINK-31646][network] Implement the remote tier producer for the tiered storage

TanYuxin-tyx opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

*Implement the remote tier producer for the tiered storage*


## Brief change log

*(for example:)*
  - *【commit 1】 Implement HashPartitionFile for the tiered storage*
  - *【commit 2】Implement the remote tier producer for the tiered storage*


## Verifying this change

This change will add tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

TanYuxin-tyx commented:
@flinkbot run azure

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierProducerAgent.java:
```diff
@@ -0,0 +1,110 @@
+        cacheDataManager.release();
+    }
+
+    private void emitBuffer(Buffer finishedBuffer, int subpartitionId) {
+        cacheDataManager.appendBuffer(finishedBuffer, subpartitionId);
+    }
```
1. The name `emit` is confusing.
2. Why do we need this method when its body is only one line of code?

TanYuxin-tyx replied:
Fixed.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierProducerAgent.java:
```diff
@@ -0,0 +1,110 @@
+
+    private final RemoteCacheManager cacheDataManager;
+
+    private final int[] currentSubpartitionWriteBuffers;
```
```suggestion
    private final int[] currentSubpartitionSegmentWriteBuffers;
```

TanYuxin-tyx replied:
Renamed.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierProducerAgent.java:
```diff
@@ -0,0 +1,110 @@
+    @Override
+    public boolean tryWrite(TieredStorageSubpartitionId subpartitionId, Buffer buffer) {
+        int subpartitionIndex = subpartitionId.getSubpartitionId();
+        if (currentSubpartitionWriteBuffers[subpartitionIndex] != 0
+                && currentSubpartitionWriteBuffers[subpartitionIndex] + 1 > numBuffersPerSegment) {
```
We already checked in the constructor that one segment at least contain one buffer. `currentSubpartitionWriteBuffers[subpartitionIndex] != 0` is probably not needed here.

TanYuxin-tyx replied:
Fixed.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java:
```diff
@@ -0,0 +1,188 @@
+
+    void startSegment(int segmentId) {
+        synchronized (allBuffers) {
+            this.segmentId = segmentId;
```
Should check `allBuffers` is empty.

TanYuxin-tyx replied:
Fixed.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java:
```diff
@@ -0,0 +1,188 @@
+            checkState(this.segmentId == segmentId, "Wrong segment id.");
+        }
+        int numFlushedBuffers = flushBuffers();
+        if (numFlushedBuffers > 0) {
+            PartitionFileWriter.SubpartitionBufferContext bufferContext =
+                    new PartitionFileWriter.SubpartitionBufferContext(
+                            subpartitionId,
+                            Collections.singletonList(
+                                    new PartitionFileWriter.SegmentBufferContext(
+                                            segmentId, Collections.emptyList(), true)));
+            // Notify the partition file writer that the segment is finished through writing the
+            // buffer context
+            flushCompletableFuture =
+                    partitionFileWriter.write(
+                            partitionId, Collections.singletonList(bufferContext));
+        }
```
This doesn't seem right. `numFlushedBuffers == 0` doesn't mean there's no data in the segment. It is possible that the buffers are flushed before calling `finishSegment`. E.g., on buffer reclaim request.

TanYuxin-tyx replied:
Fixed.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/common/TieredStorageUtils.java:
```diff
@@ -49,4 +60,61 @@ public static void setBufferWithHeader(
+                basePath, TieredStorageIdMappingUtils.convertId(partitionId), subpartitionId);
+    }
+
+    public static Path getSegmentPath(
```
`setBufferWithHeader` might be private.

TanYuxin-tyx replied:
1. Moved the methods about `segment mode` to `SegmentPartitionFile`.
2. Moved `DATA_FILE_SUFFIX` to `ProducerMergedPartitionFile`
3. Make `setBufferWithHeader` private.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/TierFactory.java:
```diff
@@ -32,7 +32,9 @@
     /** Creates the master-side agent of a Tier. */
-    TierMasterAgent createMasterAgent(TieredStorageResourceRegistry tieredStorageResourceRegistry);
+    TierMasterAgent createMasterAgent(
+            String remoteStorageBaseHomePath,
```
I think `dataFileBasePath`, `partitionFileWriter` and `partitionFileReader` shares the same problem.

TanYuxin-tyx replied:
1. Moved `remoteStorageBasePath` to the constructor.
2. `dataFileBasePath` is the result partition path, which should be an argument.
3. `partitionFileWriter` and `partitionFileReader` has been removed from the arguments. They are initialized by calling the static methods of the partition file.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/HashPartitionFile.java:
```diff
@@ -0,0 +1,35 @@
+ * The partition file in the hash mode. In this mode, each segment of one subpartition is written to
+ * an independent file.
+ */
+public class HashPartitionFile {
```
Fixed.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/HashPartitionFileWriter.java:
```diff
@@ -0,0 +1,234 @@
+ * <p>Note that after finishing writing a segment, a segment-finish file is written to ensure the
+ * downstream reads only when the entire segment file is written, avoiding partial data reads. The
+ * downstream can determine if the current segment is complete by checking for the existence of the
+ * segment-finish file.
```
Fixed. Each subpartition only keeps a segment-finish file for the latest segment.
Added doc in `SegmentPartitionFileWriter`.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/HashPartitionFileWriter.java:
```diff
@@ -0,0 +1,234 @@
+                                                subpartitionId,
+                                                segmentBuffers,
+                                                flushSuccessNotifier);
+                                ioExecutor.execute(writeRunnable);
```
Fixed.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/HashPartitionFileWriter.java:
```diff
@@ -0,0 +1,234 @@
+                                CompletableFuture<Void> flushSuccessNotifier =
+                                        new CompletableFuture<>();
+                                Runnable writeRunnable =
+                                        getFlushOrFinisheSegmentRunnable(
```
Fixed.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/HashPartitionFileWriter.java:
```diff
@@ -0,0 +1,234 @@
+                                flushSuccessNotifier)
+                : () ->
+                        writeFinishSegmentFile(
+                                partitionId, subpartitionId, segmentId, flushSuccessNotifier);
```
1. Fixed
2. `DiskCacheManager#getSubpartitionToFlushBuffers` will be fixed in the subsequential change, which is not fixed in this PR.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierMasterAgent.java:
```diff
@@ -0,0 +1,72 @@
+    @Override
+    public void release(String toRelease) {
+        deletePathQuietly(toRelease);
+    }
```
Removed.

TanYuxin-tyx commented:
@xintongsong Thanks for helping review. I addressed the comments with a fix-up commit and added the unit tests.
But the fix-up is only to be convenient for review. The fixup's files are not separated correctly. I will rebase and separate the files again after the review.

TanYuxin-tyx commented:
@flinkbot run azure

TanYuxin-tyx commented:
@flinkbot run azure

reswqa reviewed (commented):
Thanks @TanYuxin-tyx, I left some comments, please take a look.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/SegmentPartitionFile.java:
```diff
@@ -0,0 +1,159 @@
+            // Note that this check requires the file system to ensure that only one file is in the
+            // directory can be accessed when renaming a file.
+            checkState(files.length == 1, "Wong number of segment-finish files.");
+            fs.rename(files[0].getPath(), segmentFinishFile);
```
Added doc to explain this.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java:
```diff
@@ -0,0 +1,187 @@
+        synchronized (allBuffers) {
+            for (Tuple2<Buffer, Integer> bufferAndIndex : allBuffers) {
+                Buffer buffer = bufferAndIndex.f0;
+                if (!buffer.isRecycled()) {
```
Ok, simplified this method.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java:
```diff
@@ -0,0 +1,187 @@
+
+    void finishSegment(int segmentId) {
+        synchronized (allBuffers) {
+            checkState(this.segmentId == segmentId, "Wrong segment id.");
```
Removed the synchronization. Only the task thread can modify the segmentId, and the method can only be called by the task thread, so accessing segmentId is not guarded here.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/SegmentPartitionFileWriter.java:
```diff
@@ -0,0 +1,231 @@
+        int segmentId = segmentBuffers.getSegmentId();
+        List<Tuple2<Buffer, Integer>> buffersToFlush = segmentBuffers.getBufferAndIndexes();
+        boolean isSegmentFinished = segmentBuffers.isSegmentFinished();
+        checkState(!buffersToFlush.isEmpty() || isSegmentFinished);
```
This is not exclusive, so this is not changed.

TanYuxin-tyx commented:
@reswqa Thanks for helping review, I have addressed the comments, PTAL.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/SubpartitionRemoteCacheManager.java:
```diff
@@ -0,0 +1,187 @@
+    /** Release all buffers. */
+    void release() {
+        recycleBuffers();
+        checkState(allBuffers.isEmpty(), "Leaking buffers.");
```
Added the synchronization to protect this.

reswqa reviewed (approved):
Thanks for the update, LGTM.

TanYuxin-tyx commented:
@xintongsong @reswqa Thanks for helping review.
